### PR TITLE
#65: creating measure of smoothness

### PR DESCRIPTION
**Pull request by jgheerardyn**
@rombautt Could you quickly review this code  - when working with Apache Math Common Matrix objects it is a bit hard to combine with Java 8...

### CODE REVIEW

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+    @Test(expected = IllegalArgumentException.class)
+    public void constructor_orderShouldBeStrictlyPositive() {
+        new Smoothness(0);
+    }
+
+    public static class LinearData extends Functional<List<Tuple>> {
+
+        private static final double SCALE = 1e5;
+        @Variable("[-1000,1000") public double a;
```
Comment (rombautt) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: interesting, we don't need [ or ]

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+import io.yields.math.framework.property.Properties;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+@RunWith(SpecRunner.class)
+public class SmoothnessTest {
+
+    @Test(expected = IllegalArgumentException.class)
+    public void constructor_orderShouldBeStrictlyPositive() {
+        new Smoothness(0);
```
Comment (rombautt) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: also try a test with a negative and a positive value

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                return new Tuple(x, y);
+            }).collect(Collectors.toList());
+        }
+    }
+
+    @Spec(name = "linear smoothness", functional = SmoothnessTest.LinearData.class)
+    public void testLinearCase(FunctionalContext<List<Tuple>> context) {
+        //first order match should work
+        DescriptiveStatistics stats = new Smoothness(1).apply(context.evaluate());
+        assertThat(stats.getMin()).isEqualTo(stats.getMax(), Delta.delta(1.e-8)).isEqualTo(0d, Delta.delta(1.e-8));
```
Comment (rombautt) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: perfect: chaining, one of the powers of fest

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        @Variable("[-1000,1000") public double a;
+        @Variable("[-1000,1000") public double b;
+        @Variable("[-1000,1000") public double c;
+        private static final RandomSequence<Double> generator = new CommonsMathDoubleRandomSequence(new MersenneTwister(0));
+
+        @Override
+        protected List<Tuple> compute() {
+            return IntStream.rangeClosed(0, 100).mapToObj(index -> {
+                double x = (generator.next() - .5) * SCALE;
+                double y = a * x * x + b * x + c;
```
Comment (rombautt) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: I prefer to group the operators (though ofcourse not necessary) for readability: (a * x * x) + (b * x) + c;

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    @Spec(name = "quadratic smoothness", functional = SmoothnessTest.QuadraticData.class)
+    public void testQuadraticCase(FunctionalContext<List<Tuple>> context) {
+        //first order will not match
+        DescriptiveStatistics stats = new Smoothness(1).apply(context.evaluate());
+        assertThat(stats.getMin()).isGreaterThan(0);
+        assertThat(stats.getMax()).isLessThan(1);
+        assertThat(stats.getMean()).isLessThan(.5);
+        //second order matching as well
```
Comment (rombautt) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: split in 2 tests

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        @Variable("[-1000,1000") public double c;
+        @Variable("]0,100]") public double scale;
+
+        private static final RandomSequence<Double> generator = new CommonsMathDoubleRandomSequence(new MersenneTwister(0));
+
+        protected List<Tuple> sample(double scale) {
+            RandomSequence<Double> generator = new CommonsMathDoubleRandomSequence(new MersenneTwister(0));
+            return IntStream.rangeClosed(0, 100).mapToObj(index -> {
+                double x = (generator.next() - .5) * SCALE;
+                double y = (a * x * x + b * x + c) * scale;
```
Comment (rombautt) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: same comment as above

File: concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        return stats;
+    }
+
+    private RealMatrix computeDistance(List<Tuple> normalizedData) {
+        RealMatrix a = new Array2DRowRealMatrix(normalizedData.size(), order + 1);
+        RealMatrix b = new Array2DRowRealMatrix(normalizedData.size(), 1);
+        int row = 0;
+        for (Tuple tuple : normalizedData) {
+            final int currentRow = row;
```
Comment (rombautt) [inline concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java]: what's the added value of this variable?

File: concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Override
+    public DescriptiveStatistics apply(Collection<Tuple> tuples) {
+        //first we normalize the tuples so data fits in the unit square
+        List<Tuple> normalizedData = normalize(tuples);
+        //calculate error (i.e. distance between tuple y and fitted polynomial
+        RealMatrix error = computeDistance(normalizedData);
+        //group in stats object
+        DescriptiveStatistics stats = new DescriptiveStatistics();
```
Comment (rombautt) [inline concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java]: Java 8 has a collector builtin to get the statistics of a collection, as you only use min, max and mean, this might be a better fit:

error.getColumn(0).stream().collect(Collectors.summarizingDouble), gives a DoubleSummaryStatistics.

http://docs.oracle.com/javase/8/docs/api/java/util/DoubleSummaryStatistics.html

**Comment by rombautt** [comment]
reviewed

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java (hunk 1 above)
Comment (jgheerardyn) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: robust parsing indeed :-)

File: concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java (hunk 2 above)
Comment (jgheerardyn) [inline concepts/src/test/java/io/yields/math/concepts/operator/SmoothnessTest.java]: added strictly negative value - positive values are used in other tests. Don't see what additional added value a test with a positive value would bring

File: concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java (hunk 7 above)
Comment (jgheerardyn) [inline concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java]: lambda expression requires a final - I agree it is not clean but don't see how else to do it...

File: concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java (hunk 8 above)
Comment (jgheerardyn) [inline concepts/src/main/java/io/yields/math/concepts/operator/Smoothness.java]: nice
